init function is fallible

Wenzel opened the pull request:
This PR should help solve #7 

- dummy driver has been removed
- add `enum-iterator` to iterate over the `DriverType` enum
- init function is now fallible
- all driver init functions are now fallible

Currently when you try to execute an example on Xen, it will panic in the KVM driver, as it is initialized before KVM:
![Capture d’écran de 2020-12-30 18-06-15](https://user-images.githubusercontent.com/964610/103369185-c2ae0b00-4ac9-11eb-8b8f-2cec94d5db95.png)

Wenzel commented:
@rageagainsthepc i've updated this PR, and removed setting all 3 linux drivers as default features in the build system.
It's not the behavior that I wanted.

I redesigned the init function here, to make it fallible, and to iterate over each compiled driver, to return the first one that succeeds.

I tried to create a Result for the C API interface, it compiles, but cbindgen doesn't generate a valid libmicrovmi header.
Please give me your opinion on this PR when you have time

rageagainsthepc commented on src/capi.rs:
```diff
@@ -44,8 +52,10 @@ pub unsafe extern "C" fn microvmi_init(
     };
-    let driver = init(&safe_domain_name, optional_driver_type, init_option);
-    Box::into_raw(Box::new(driver)) as *mut c_void
+    match init(&safe_domain_name, optional_driver_type, init_option) {
```
Ok, so cbindgen obviously has problems converting complex enums into C structs, but that shouldn't be surprising because there is no real counterpart for those in C. Maybe a union type wrapping multiple structs. However I would not try to replicate rust error handling in C because idiomatic error handling in C just works differently. I would simply return null in error cases and additionally add an out parameter so we can set error values. The error value could be a simply enum or maybe a struct containing an enum and a string description of the error.

rageagainsthepc commented on src/lib.rs:
```diff
@@ -19,67 +21,57 @@ use driver::kvm::Kvm;
+                        return Ok(driver);
+                    }
+                    Err(e) => {
+                        info!("{}", e);
```
nit: I would not display DriverNotCompiled errors on info log level.

Wenzel commented:
modifications:

- updated the init C function to take an additional `**const c_char`, to set a string message in case of microvmi init failure
- changed the last MicrovmiError variant display message to display its source

and it works ! :tada: 
![Capture d’écran de 2021-01-31 18-04-04](https://user-images.githubusercontent.com/964610/106392060-90593b80-63f0-11eb-85b2-ed5b27e4be8a.png)


C API
![Capture d’écran de 2021-01-31 18-15-26](https://user-images.githubusercontent.com/964610/106392068-93ecc280-63f0-11eb-9953-b27d60ca63de.png)

Wenzel commented on src/capi.rs:
```diff
@@ -23,12 +24,18 @@ pub unsafe extern "C" fn microvmi_envlogger_init() {
     domain_name: *const c_char,
     driver_type: *const DriverType,
     driver_init_option: *const DriverInitParamFFI,
+    init_error: *mut *const c_char,
```
if the initialization fails, I don't see what the user can do except log and exit.
of course it would be better to give more meaningful info to the C world, what type of error is reported for example.

libvmi fills an enum to try to convey the cause of the failure.
https://github.com/libvmi/libvmi/blob/master/libvmi/libvmi.h#L93

libmicrovmi is already doing better by providing a precise error string, and not a generic one i'd say.

Wenzel replied:
f that's okay for you, you can resolve this conversation.

rageagainsthepc commented on src/capi.rs:
```diff
@@ -17,18 +18,24 @@ pub enum DriverInitParamFFI {
+///
+/// In case of error, init_error will be allocated with the underlying error message.
+/// The init_error pointer should be freed with rs_cstring_free()
+/// # Safety
```
Is that an empty Safety section? Why?

Wenzel replied:
because there is no additional safety comment to add, but clippy will throw a warning without this section.
maybe i should have disabled the clippy warning then ?

rageagainsthepc replied:
Well, the safety section should document the caller's responsibilities in order to uphold safety guarantees, so you should move line 31 into this section, for example.
